Benchmarks/GaussianElimination.cs

```diff
@@ -0,0 +1,278 @@
+                    if (pi != j)
+                    {
+                        var tmp = Ab[pi];
+                        Ab[pi] = Ab[j];
+                        Ab[j] = tmp;
+                    }
+
+                    // Eliminate the rows after the pivot.
+                    double p = Ab[j][j];
+                    for (int i = j + 1; i < M; ++i)
+                    {
+                        double s = Ab[i][j] / p;
+                        if (s != 0.0d)
+                        {
+                            for (int jj = j + 1; jj <= N; ++jj)
+                            {
+                                Ab[i][jj] -= Ab[j][jj] * s;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        [Benchmark]
+        public void ArrayOfArraysVectorized()
+        {
+            for (int iteration = 0; iteration < Size; iteration++)
+            {
+                var Ab = _data[iteration].JaggedArray;
+
+                const double tiny = 0.00001;
+
+                // Solve for dx.
+                // For each variable in the system...
+                for (int j = 0; j + 1 < N; ++j)
+                {
+                    int pi = j;
+                    double max = Math.Abs(Ab[j][j]);
+
+                    // Find a pivot row for this variable.
+                    for (int i = j + 1; i < M; ++i)
+                    {
+                        // if(|JxF[i][j]| > max) { pi = i, max = |JxF[i][j]| }
+                        double maxj = Math.Abs(Ab[i][j]);
+                        if (maxj > max)
+                        {
+                            pi = i;
+                            max = maxj;
+                        }
+                    }
+
+                    // Swap pivot row with the current row.
+                    if (pi != j)
+                    {
+                        var tmp = Ab[pi];
+                        Ab[pi] = Ab[j];
+                        Ab[j] = tmp;
+                    }
+
+                    var vectorLength = System.Numerics.Vector<double>.Count;
+                    // Eliminate the rows after the pivot.
+                    double p = Ab[j][j];
+                    for (int i = j + 1; i < M; ++i)
+                    {
+                        double s = Ab[i][j] / p;
+                        if (Math.Abs(s) >= tiny)
+                        {
+                            int jj;
+                            for (jj = j + 1; jj <= (N - vectorLength); jj += vectorLength)
+                            {
+                                var source = new System.Numerics.Vector<double>(Ab[j], jj);
+                                var target = new System.Numerics.Vector<double>(Ab[i], jj);
+                                var res = target - (source * s);
+                                res.CopyTo(Ab[i], jj);
+                            }
+                            for (; jj <= N; ++jj)
```

Comment:
It would be fine to pad the matrix to a multiple of the vector size so we don't need a scalar loop tail. Given the sizes of matrices involved for us, this might make a big difference in performance.

Reply:
The tail loop is here just in case, to make the function more robust (benchmark with M=12 doesn't use it, as on most processors vector of doubles have a length of 4), but for the actual implementation we can just pad. 😉

Reply:
I'm wondering if the extra loop overhead might affect the benchmark though? The main loop might just be 2 loads + fma + store, and so if there is only one or two iterations of that loop, the extra loop overhead for the scalar tail might be significant?